### DISCUSSION TITLE
Difference in handling RaiseException in psycopg2 vs psycopg3 drivers

### DISCUSSION THREAD

**Discussion by lelit** [sqlalchemy/sqlalchemy#12144] [Usage Questions] [answered]
Hi,

in one of my applications that uses `PostgreSQL`, I have a set con _constraint triggers_ that checks for some condition, and when they are not met they uses the [RAISE EXCEPTION](https://www.postgresql.org/docs/17/plpgsql-errors-and-messages.html) statement.

Trying to switch from `psycopg2` to `psycopg` (that is, 3.x), I notice a change in how SA handles the error returned by the driver: both morph the PG error into a `RaiseException` exception, but while the `psycopg2` SA driver wraps it in a `sqlalchemy.exc.InternalError`, the `psycopg3` one wraps it in a `sqlalchemy.exc.ProgrammingError`.

The following sample script exhibits the difference:
```python
from psycopg2 import connect as psycopg2_connect
from psycopg import connect as psycopg3_connect

from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy import create_engine


FUNC = """
CREATE OR REPLACE FUNCTION error(error_id integer)
RETURNS void AS $$

DECLARE
  info text;

BEGIN
  info = '(' || error_id || ')';

  RAISE EXCEPTION USING ERRCODE='P0001', MESSAGE='error', HINT=info;
END
$$ LANGUAGE plpgsql STABLE
"""

print("PSYCOPG2")

psycopg2_c = psycopg2_connect('dbname=test')

with psycopg2_c.cursor() as c:
    c.execute(FUNC)

with psycopg2_c.cursor() as c:
    try:
        c.execute('SELECT error(2)')
    except Exception as e:
        print(f'Got a {e.__class__} exception: {e}')

print()


print("PSYCOPG3")

psycopg3_c = psycopg3_connect('dbname=test')

with psycopg3_c.cursor() as c:
    try:
        c.execute('SELECT error(3)')
    except Exception as e:
        print(f'Got a {e.__class__} exception: {e}')

print()


print("SA+PSYCOPG2")

psycopg2_e = create_engine("postgresql+psycopg2://localhost/test")

with psycopg2_e.connect() as conn:
    try:
        conn.execute(select(func.error(2)))
    except Exception as e:
        print(f'Got a {e.__class__} exception: {e}')

print()

print("SA+PSYCOPG3")

psycopg3_e = create_engine("postgresql+psycopg://localhost/test")

with psycopg3_e.connect() as conn:
    try:
        conn.execute(select(func.error(3)))
    except Exception as e:
        print(f'Got a {e.__class__} exception: {e}')
```

It's output is the following:

```
PSYCOPG2
Got a <class 'psycopg2.errors.RaiseException'> exception: error
HINT:  (2)
CONTEXT:  PL/pgSQL function error(integer) line 9 at RAISE


PSYCOPG3
Got a <class 'psycopg.errors.RaiseException'> exception: error
HINT:  (3)
CONTEXT:  PL/pgSQL function error(integer) line 9 at RAISE

SA+PSYCOPG2
Got a <class 'sqlalchemy.exc.InternalError'> exception: (psycopg2.errors.RaiseException) error
HINT:  (2)
CONTEXT:  PL/pgSQL function error(integer) line 9 at RAISE

[SQL: SELECT error(%(error_2)s) AS error_1]
[parameters: {'error_2': 2}]
(Background on this error at: https://sqlalche.me/e/20/2j85)

SA+PSYCOPG3
Got a <class 'sqlalchemy.exc.ProgrammingError'> exception: (psycopg.errors.RaiseException) error
HINT:  (3)
CONTEXT:  PL/pgSQL function error(integer) line 9 at RAISE
[SQL: SELECT error(%(error_2)s::INTEGER) AS error_1]
[parameters: {'error_2': 3}]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

Is that intentional, or something that deserves further investigation and maybe a fix?

Thanks in advance,
bye, lele.

**Comment by zzzeek** [comment, marked as answer]
that's the driver.  SQLAlchemy does that wrapping based on what the driver raises and nothing more.

**Reply by lelit** [reply]
Thank you, I wrongly assumed that the mapping were done in SA.

So, just to be precise, [psycopg v2](https://github.com/psycopg/psycopg2/blob/master/psycopg/error_type.c#L132) maps `P0001` [error code](https://www.postgresql.org/docs/current/errcodes-appendix.html) to a subclass of `InternalError`, while [psycopg v3](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/errors.py#L1718) maps it to a subclass of `ProgrammingError`. Both ways are somewhat less than intuitive for a reason or another, but that's the way it is.

I will ask Daniele an opinion.
